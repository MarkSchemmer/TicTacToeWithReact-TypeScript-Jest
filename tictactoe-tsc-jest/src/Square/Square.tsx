import * as React from 'react'

interface IProps {
    coordinates : Array<number>,
    val : string | null 
}


class Square extends React.Component<IProps> {
    render() {
    const { coordinates, val } = this.props 
        return (
            <div className="square">
                Square: { coordinates } { val }
            </div>
        )
    }
}


export default Square



