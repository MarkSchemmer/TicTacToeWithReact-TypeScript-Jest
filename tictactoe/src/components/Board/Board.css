
.row { 
    width:400px;
    height: 100px;
    /* border: 1px solid black; */
    margin: auto;
}


.column { 
    width: 130px;
    height: 100px;
    border : 1px solid black;
    display: inline-block;
}