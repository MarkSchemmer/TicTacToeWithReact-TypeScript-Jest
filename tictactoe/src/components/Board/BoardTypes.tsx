import React, { Component } from 'react'


interface ISquareAttributes {
    id : number, 
    coor : Array<number> 
    val : string | null 
}

interface IBoardWithMove {
    history : Array<Array<ISquareAttributes>>,
    Move : Array<Array<number>> | null,
    WhomMoved : string | null 
}


export class SquareAttributes implements ISquareAttributes {
    public id : number; 
    public coor : Array<number>;
    public val : string | null;

    constructor(id : number, coor : Array<number>, val : string | null = null){
        this.id = id
        this.coor = coor 
        this.val = val 
    }
}

export class BoardWithMoves implements IBoardWithMove {
    
    public history : Array<Array<SquareAttributes>>

    public Move : Array<Array<number>>

    public WhomMoved : string | null 
    
    constructor(history:Array<Array<SquareAttributes>>,
         Move : Array<Array<number>>, 
         WhomMoved : string | null = null ){
        this.history = history
        this.Move = Move 
        this.WhomMoved = WhomMoved
    }
}