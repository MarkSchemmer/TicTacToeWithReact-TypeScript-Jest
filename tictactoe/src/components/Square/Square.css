
span {
}

span div {
    float : left;
    width: 100%;
    height: 50%;
    padding-top:40px;
    margin : auto;
    font-size : 2em;
    font-weight: bolder;
}