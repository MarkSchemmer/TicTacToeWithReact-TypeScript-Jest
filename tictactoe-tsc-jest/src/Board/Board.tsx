import * as React from 'react'
import Square from '../Square/Square'
import * as _ from '../Helpers/helpers'
interface IProps  {
    squares : Array<Array<_.Square>>
}

class Board extends React.Component<IProps> {
    render() {
        const { squares } = this.props
        return (
            <div>
                { squares.map((level) => level.map((sublevel) => {
                        return <Square
                                val={sublevel.val} 
                                coordinates={sublevel.coordinates}  /> 
                }))}          
            </div>
        )
    }
}

export default Board 