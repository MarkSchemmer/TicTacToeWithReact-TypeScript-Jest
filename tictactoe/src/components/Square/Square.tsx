import React, { Component } from 'react'
import './Square.css'

interface IProps {
    id : number, 
    coor: Array<number>,
    val : string | null,
    onClick(coor:Array<number>) : void 
}

interface IState { }

class Square extends Component<IProps, IState> {


    
    constructor( props : IProps){
        super(props)
    }


    public render () {
        return (
            <React.Fragment>
                <span onClick={() => this.props.onClick(this.props.coor)} className="column"> 
                       <div className="square-val">{ this.props.val }</div> 
                </span>
            </React.Fragment>
        )
    }
}

export default Square 

