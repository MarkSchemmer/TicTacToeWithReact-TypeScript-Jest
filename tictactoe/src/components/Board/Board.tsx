import React, { Component } from 'react'
import Square from '../Square/Square'
import './Board.css'


interface IProps {

}

interface IState {
    name : string, 
    board : any,
    turn : number 
}

class Board extends Component<IProps, IState> {


    public state : IState = {
        name : 'Mark',
        board : genBoard(),
        turn : 0
    }


    public onClick = (coor:Array<number>) => {
        const [x,y] = coor
        if(this.state.board[x][y].val !== null) return  
        let copy = Object.assign({}, 
            { board : this.state.board, turn : this.state.turn+1, name: 'Mark' })
        if(this.state.turn%2==0)
            copy.board[x][y].val = 'X'
        else 
            copy.board[x][y].val = 'O'
        this.setState(copy)
    }



    public render () {
        return (
            <React.Fragment>
                <div className="main">
                  <div className="greetings">Greetings { this.state.name }</div>
                  <div className="whoms-turn"> Player { this.state.turn % 2 === 0 ? <strong>{'X'}</strong> : <strong>{'O'}</strong> } your move! </div>
                    <div className="board">
                            { this.state.board.map((x:any) => 
                               <div className="row" key={Date.now()*(Math.random()*100)}>
                                    { x.map((z:any) => <Square key={z.id} {...z} 
                                        onClick={this.onClick.bind(this)} /> )}
                               </div> )}
                    </div>
                </div>
            </React.Fragment>
        )
    }
}

export default Board 


//let genBoard = () => [...Array(3)].map(x => [...Array(3)].map( y => ({ val: null, coor : [x,y]  }) ))
let id = 0
let genBoard = () => Array.from(new Array(3), (val, x) => {
    return Array.from(new Array(3), (otherVal, y) => ({
        id : id++,
        coor : [x,y], 
        val : null 
    }))
} )
