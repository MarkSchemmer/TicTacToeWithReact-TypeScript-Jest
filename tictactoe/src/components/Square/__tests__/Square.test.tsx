import * as React from 'react'
import * as enzyme from 'enzyme'
import  Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import Board from '../../Board/Board'
import Square from '../Square'

let fakeProps = {
    id : 1, 
    coor: [0,0],
    val : null 
}
 
enzyme.configure({ adapter: new Adapter() });

describe('dummy test for Square ', () => {

    it('dummy', () => {
        expect(true).toEqual(true)
    })

})

describe(' Simulating events on <Square />', () => {

    it(' Will onClick be fired when clicked ', () => { 
        const onClickSpy = jest.fn()
        const wrapper = enzyme.shallow( <Square {...fakeProps} onClick={onClickSpy} /> )
        wrapper.find('.column').simulate('click')
        expect(onClickSpy).toHaveBeenCalled() 
    })

    it('render <Board /> and simulate change on first <Square /> ', () => {
        const onClickSpy = jest.fn()
        const wrapper = enzyme.shallow( <Square {...fakeProps} onClick={onClickSpy} /> )

        wrapper.find('.column').simulate('click')

        expect(onClickSpy).toHaveBeenCalled()
        
        wrapper.setProps({ val : 'X' })

        expect(wrapper.find('.square-val').text()).toBe('X')
    })

    it(' test update for <Square /> component ', () => {

        const onClickSpy = jest.fn()

        const wrapper = enzyme.mount( <Board /> )

        expect(wrapper.find('.column')).toHaveLength(9)

        wrapper.find('.column').at(0).simulate('click')

        wrapper.update() 

        expect( wrapper.find('.column').at(0).text() ).toEqual('X')

    })
})

describe(' snapshot testing for <Square /> ', () => {
    it('snapshot of square... ', () => {
        const tree = renderer.create(<Square {...fakeProps} onClick={jest.fn()} /> ).toJSON()
        expect(tree).toMatchSnapshot() 
    })
})