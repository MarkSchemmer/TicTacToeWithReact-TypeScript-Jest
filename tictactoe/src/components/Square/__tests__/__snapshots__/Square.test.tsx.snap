// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` snapshot testing for <Square />  snapshot of square...  1`] = `
<span
  className="column"
  onClick={[Function]}
>
  <div
    className="square-val"
  />
</span>
`;
